cpp/include/cudf/io/text/data_chunk_source.hpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+#include <rmm/device_buffer.hpp>
+
+namespace cudf {
+namespace io {
+namespace text {
+
+/**
+ * @brief a reader capable of producing views over device memory.
+ *
+ * The data chunk reader API encapsulates the idea of statefully traversing and loading a data
+ * source. A data source may be a file, a region of device memory, or a region of host memory.
+ * Reading data from these data sources efficiently requires different strategies dependings on the
+ * type of data source, type of compression, capabilities of the host and device, the data's
+ * destination. Whole-file decompression should be hidden behind this interface
+ *
+ */
+class data_chunk_reader {
+ public:
+  /**
+   * @brief Get the next chunk of bytes from the data source
+   *
+   * Performs any necessary work to read and prepare the underlying data source for consumption as a
+   * view over device memory. Common implementations may read from a file, copy data from host
+   * memory, allocate temporary memory, perform iterative decompression, or even launch device
+   * kernels.
+   *
+   * @param size desired number of bytes
+   * @param stream stream to associate allocations or perform work required to obtain chunk
+   * @return a chunk of data up to @param size bytes, or less if no more data is avaialable
+   */
+  virtual device_span<char const> get_next_chunk(uint32_t size, rmm::cuda_stream_view stream) = 0;
+};
+
+/**
+ * @brief a data source capable of creating a reader which can produce views of the data source in
+ * device memory.
+ *
+ */
+class data_chunk_source {
```

Comment:
I don't understand the purpose of this polymorphic class in addition to `data_chunk_reader`.

Why can't the interface to `multibyte_split` be:
```
std::unique_ptr<cudf::column> multibyte_split(
  data_chunk_reader const& source,
  std::string const& delimiter,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

All that I can see that is done with the `data_chunk_source` is call `create_reader()`  which returns a polymorphic `data_chunk_reader`. This seems like an unnecessary step that could be simplified by just accepting a `data_chunk_reader` to begin with.

What am I missing?

Reply:
I want to use a forward_iterator-like class to enforce the idea that each pass is a sequential read. `data_chunk_reader` provides this functionality - it maintains the read position. I also need the ability to run two passes, so `data_chunk_source` allows creating multiple forward_iterator-like classes. In the future, `data_chunk_source` will evolve in to a random-access data source, and `data_chunk_reader` will remain forward-only.

Both classes will go through more design iterations before being adopted by the rest of cuio.